lib/object_pool.js

```diff
@@ -0,0 +1,171 @@
+ObjectPool.timer = null;
+ObjectPool.refs = 0;
+
+ObjectPool.setup = function setup(options) {
+    var pool = new ObjectPool(options);
+    options.Type.alloc = alloc;
+
+    options.Type.prototype.free = function freeThisObj() {
+        pool.free(this);
+    };
+
+    ObjectPool.pools.push(pool);
+
+    return pool;
+
+    function alloc() {
+        var obj = pool.get();
+        return obj;
+    }
+};
+
+ObjectPool.bootstrap = function bootstrap(options) {
+    if (ObjectPool.refs >= 1) {
+        ObjectPool.refs += 1;
+        return;
+    }
+
+    ObjectPool.refs += 1;
+
+    assert(typeof options === 'object', 'expected options object');
+
+    assert(
+        typeof options.statsd === 'object' &&
+        typeof options.statsd.gauge === 'function',
+        'expected options.statsd to be statsd instance'
+    );
+    ObjectPool.statsd = options.statsd;
+
+    assert(
+        typeof options.reportInterval === 'number',
+        'expected options.reportInterval to be number'
+    );
+    ObjectPool.reportInterval = options.reportInterval;
+
+    assert(
+        typeof options.timers === 'object' &&
+        typeof options.timers.setTimeout === 'function',
+        'expected options.timers to be timers object'
+    );
+    ObjectPool.timers = options.timers;
+
+    ObjectPool.timer = ObjectPool.timers.setTimeout(
+        ObjectPool.reportStats,
+        ObjectPool.reportInterval
+    );
+};
+
+ObjectPool.unref = function unref() {
+    ObjectPool.refs = Math.min(0, ObjectPool.refs - 1);
+    if (ObjectPool.refs === 0) {
+        ObjectPool.timers.clearTimeout(ObjectPool.timer);
+    }
+};
+
+ObjectPool.reportStats = function reportStats() {
+    // Iterate over pools
+    // report their current size
+    var i;
+    var pool;
+    for (i = 0; i < ObjectPool.pools.length; i++) {
+        pool = ObjectPool.pools[i];
+        if (ObjectPool.statsd) {
+            ObjectPool.statsd.gauge(
+                'object-pools.' + pool.name + '.free',
+                pool.freeList.length
+            );
+
+            // This stat is how we infer the maxSize property
+            ObjectPool.statsd.gauge(
+                'object-pools.' + pool.name + '.outstanding',
+                pool.outstanding
+            );
+        }
+    }
+
+    ObjectPool.timer = ObjectPool.timers.setTimeout(
+        ObjectPool.reportStats,
+        ObjectPool.reportInterval
+    );
+};
+
+ObjectPool.prototype.get = function get() {
+    var inst;
+    this.outstanding += 1;
+    if (this.freeList.length) {
+        inst = this.freeList.pop();
+        assert(inst._objectPoolIsFreed, 'instance retreived from pool is free');
+        inst._objectPoolIsFreed = false;
+        return inst;
+    } else {
+        inst = new this.Type();
+        inst._objectPoolIsFreed = false;
+        return inst;
+    }
+};
+
+ObjectPool.prototype.free = function free(inst) {
+    assert(!inst._objectPoolIsFreed, 'object pool double free');
+    inst._objectPoolIsFreed = true;
```

Comment:
It does, but since *all* objects that are of a class setup to be pooled will have this property, it shouldn't have an affect on optimization since the functions will all be compiled against the map for `LazyRelayInReq'`.